src/backend/pipeline/fss.c

```diff
@@ -87,12 +107,63 @@ FSSDestroy(FSS *fss)
 	pfree(fss);
 }
 
+/*
+ * fss_resize_topk
+ *
+ * Resize the given FSS so that it can contiguously store the given top k array
+ */
+static FSS *
+fss_resize_topk(FSS *fss, ArrayType *top_k)
+{
+	Size delta = Abs((int) ARR_SIZE(top_k) - (int) ARR_SIZE(fss->top_k));
+	FSS *result = fss;
+
+	if (delta != 0)
+	{
+		result = repalloc(result, FSSSize(fss) + delta);
+		result = FSSFromBytes((struct varlena *) result);
+	}
+	memcpy(result->top_k, top_k, ARR_SIZE(top_k));
+	SET_VARSIZE(result, FSSSize(result));
+
+	return result;
+}
+
+/*
+ * get_monitored_value
+ *
+ * Get the monitored value, which could be either byval or byref. If it's byref,
+ * we need to pull the actual value out of FSS's varlena array.
+ */
+static Datum
+get_monitored_value(FSS *fss, MonitoredElement *element, bool *isnull)
+{
+	Datum result = element->value;
+	int index;
+
+	*isnull = false;
+
+	if (fss->typ.typbyval)
```

Comment:
What about null check for by val types? Seems like we'll always return `isnull = false`.